core/src/validator.rs

```diff
@@ -24,234 +17,217 @@ pub enum Error {
+        Result::<(), data_model_validator::DenialReason>::from(verdict).map_err(|reason| {
+            Error::ValidatorDeny {
+                operation: operation.clone(),
+                reason,
+            }
         })
     }
 }
 
-/// Constant view to the [`Chain`].
-///
-/// Provides [`Chain`] const methods without interior mutability.
+#[cfg(test)]
 #[derive(Debug, Copy, Clone)]
-pub struct ChainView<'chain> {
-    chain: &'chain Chain,
-}
+pub struct MockValidator;
 
-impl ChainView<'_> {
-    /// Wrapper around [`Self::validate()`].
+#[cfg(test)]
+impl MockValidator {
+    /// Mock for operation validation.
     ///
     /// # Errors
-    /// See [`Chain::validate()`].
+    ///
+    /// Never fails.
+    #[allow(
+        clippy::unused_self,
+        clippy::unnecessary_wraps,
+        clippy::trivially_copy_pass_by_ref,
+        clippy::needless_pass_by_value
+    )]
     pub fn validate(
-        self,
-        wsv: &WorldStateView,
-        authority: &<Account as Identifiable>::Id,
-        operation: impl Into<NeedsPermissionBox>,
+        &self,
+        _wsv: &WorldStateView,
+        _authority: &<Account as Identifiable>::Id,
+        _operation: impl Into<data_model_validator::NeedsPermissionBox>,
     ) -> Result<()> {
-        self.chain.validate(wsv, authority, operation)
+        Ok(())
+    }
+}
+
+/// [`Validator`] with [`Module`](wasmtime::Module) for execution.
+///
+/// Creating [`Module`] is expensive, so we do it once on [`upgrade()`](Validator::upgrade)
+/// step and reuse it on [`validate()`](Validator::validate) step.
+#[derive(Clone)]
+struct LoadedValidator {
+    #[cfg_attr(test, allow(dead_code))]
+    module: wasmtime::Module,
+}
+
+impl Debug for LoadedValidator {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.debug_struct("LoadedValidator")
+            .field("module", &"<Module is truncated>")
+            .finish()
+    }
+}
+
+impl LoadedValidator {
+    pub fn load(
+        engine: &wasmtime::Engine,
+        raw_validator: data_model_validator::Validator,
+    ) -> Result<Self> {
+        Ok(Self {
+            module: wasm::load_module(engine, raw_validator.wasm)?,
+        })
+    }
+}
+
+/// Constant view to a [`Validator`] used by [`WorldStateView`].
+///
+/// Serves to the same purpose as [`RwLockReadGuard`](parking_lot::RwLockReadGuard),
+/// but holds [`Option`] instead of [`Validator`].
+/// That is required because [`WorldStateView`] may have uninitialized [`Validator`].
+/// However we still want to provide direct access to [`Validator`] for users, so that they
+/// don't have to deal with [`Option`].
+///
+/// # Panic
+///
+/// That said, [`new()`](Self::new) and [`deref()`](std::ops::Deref::deref) will panic if option is [`None`].
+#[derive(Debug)]
+pub struct View<'validator>(
+    #[cfg(not(test))] parking_lot::RwLockReadGuard<'validator, Option<Validator>>,
+    #[cfg(test)] MockValidator,
+    #[cfg(test)] std::marker::PhantomData<&'validator ()>,
```

Comment:
```suggestion
    #[cfg(test)] parking_lot::RwLockReadGuard<'validator, Option<MockValidator>>,
```

how about this? It could be easier to reason about and would be closer to the real system performance